x/meshsecurity/keeper/stake.go

```diff
@@ -59,12 +60,57 @@ func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.Val
 	// and update our records
 	k.setTotalDelegatedAmount(cacheCtx, actor, newTotalDelegatedAmount)
 	done()
-
-	// TODO: emit events?
-	// TODO: add to telemetry?
 	return newShares, err
 }
 
+// Undelegate executes an instant undelegate and burns the released virtual staking tokens.
+// The amount burned is added to the (negative) SupplyOffset, when supported.
+// Authorization of the actor should be handled before entering this method.
 func (k Keeper) Undelegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) error {
-	panic("not implemented, yet")
+	if amt.Amount.IsNil() || amt.Amount.IsZero() || amt.Amount.IsNegative() {
+		return errors.ErrInvalidRequest.Wrap("amount")
+	}
+
+	// Ensure staking constraints
+	bondDenom := k.staking.BondDenom(pCtx)
+	if amt.Denom != bondDenom {
+		return errors.ErrInvalidRequest.Wrapf("invalid coin denomination: got %s, expected %s", amt.Denom, bondDenom)
+	}
+
+	cacheCtx, done := pCtx.CacheContext() // work in a cached store (safety net?)
+	totalDelegatedAmount := k.getTotalDelegatedAmount(cacheCtx, actor)
+	if amt.Amount.GT(totalDelegatedAmount) {
+		return errors.ErrInvalidRequest.Wrap("amount exceeds total delegated")
+	}
+	shares, err := k.staking.ValidateUnbondAmount(cacheCtx, actor, valAddr, amt.Amount)
+	if err == stakingtypes.ErrNoDelegation {
```

Comment:
Question, when this happens, doesn't it mean we aren't able to actually unbond?
I see returning early, but why no error here?

Reply:
There is either nothing to unbond anymore or there was never a delegation to this validator. 
You can argue that this is a noop, now that does not hurt. The code comes from the Osmosis SDK fork. In order to stay compatible, we should have the same behaviour.